### ISSUE TITLE
Bug: [require-await] fails to infer types for function declarations implementing interfaces

### ISSUE THREAD

**Issue by FauxFaux** [typescript-eslint/typescript-eslint#11085]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsPhAI4Cuc1JAhgO7NxIBcYARBPnwB7fLzAhgAKEmIkAgGbNcEMADEhQsAG9JYMPIQ8AFAEowAXgB8YAArCAtnCgQAPADchcACaWA3JIBfaXk6BFwkOCEEMHtmAGsIUx51TR09XCioFHkNZI0LbV09fUMwZigMMLBTC2s04vTMoRgIADoYIQBzIwByAAsIGA6wVhEYLx6TfwaAgBoigOmwaiQ6fGicoX8goA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
/* eslint @typescript-eslint/require-await: "error" */

interface Foo {
  fn: () => Promise<void>;
}

function make(): Foo {
  const foo: Foo = {
    fn: async () => {
      console.log('hello world');
    },
  };
  return foo;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The rule believes that the `fn` implementation inside the interface must not be `async`, as it lacks an `await` expression or a `return Promise..`. However, making it not `async` causes the code to fail to compile: it must be `async` according to the types. I believe the purpose of this rule (over the eslint one) is to use this type information to not flag this code as incorrect.

### Actual Result

```
Async method 'fn' has no 'await' expression. 9:5 - 9:15
> Remove 'async'.
```

**Comment by FauxFaux** [comment]
Okay and now I find this much older discussion. I strongly disagree, but if that is still the decision then I understand.

https://github.com/typescript-eslint/typescript-eslint/issues/7450#issuecomment-1694055964

**Comment by bradzacher** [comment]
This is working as expected -- if you're using `async` as a side-effect to make a sync function return a promise then you should use a disable comment.
